### PR TITLE
feat(memory): add persistent cross-session memory capability

### PR DESCRIPTION
**Pull request by chaliy**
## Summary

- Add persistent cross-session memory capability with multi-store support per org (selected via capability config)
- Implement `remember`, `recall`, `forget` tools with capacity limits (2K char content, 4 images/memory, 10K memories/store, 50 stores/org)
- Recall returns multicontent `MemoryContentPart` (text + image variants, same shape as message `ContentPart`)
- Add `MemoryStoreBackend` trait, `InMemoryMemoryStore` impl, typed IDs (`mst_`/`mem_`), and full ToolContext wiring
- Add "Knowledge Base Agent" seed example and `specs/memory.md` specification

## Review comments addressed

All 11 Copilot reviewer comments addressed in second commit:
- **Tenant isolation**: `forget()` now requires `store_id`; `resolve_store_id` verifies store exists and belongs to caller's org before returning
- **Input validation**: Invalid `kind` values now return `ToolError` instead of silently defaulting (both `remember` and `recall`)
- **Spec accuracy**: Fixed store ID prefix (`mst_`), added `images` to remember schema, fixed recall image shape (`base64`/`media_type` not `url`), corrected capacity wording (hard cap not auto-archive), removed dedup claims (create-only for V1)
- **Debug impl**: Added `memory_store`/`org_id` to `ToolContext` Debug output
- **Config resolution**: Documented as V1 TODO — requires ToolContext extension for per-agent capability config

## Security

Threat categories reviewed: **TM-TENANT**, **TM-API**, **TM-TOOL**

- **TM-TENANT**: `resolve_store_id` now verifies `store.org_id == context.org_id` before allowing access to a configured store. `forget` passes `store_id` to backend so implementations enforce memory-to-store ownership. Error messages use generic "not found" to avoid leaking cross-org store existence.
- **TM-API**: All tool inputs validated — content length, tag count, image count/size, kind enum values. Invalid kind returns explicit error. Memory ID parsing validates `mem_` prefix.
- **TM-TOOL**: Tools require context (`requires_context() = true`), fail cleanly without memory store or org_id. No new external service calls or sandbox escapes.

No new threats introduced. No changes to `specs/threat-model.md` needed (memory stores are a new internal capability, not a new external trust boundary).

## Test plan

- [x] 15 unit tests covering remember/recall/forget, image attachments, capacity limit validation, kind filtering, edge cases
- [x] All core lib tests pass (including updated capability registry tests)
- [x] Server seed agent tests pass (unique slugs, find by slug)
- [x] `cargo fmt --check` clean
- [x] `cargo clippy --all-targets --all-features -- -D warnings` clean
- [ ] CI green

### CODE REVIEW

File: specs/memory.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+## Capability Config
+
+```json
+{
+  "ref": "memory",
+  "config": {
+    "store": "memstore_abc123",
+    "passive_recall_count": 5
+  }
```
[the remark this answers is not in the document]
Comment (chaliy) [reply specs/memory.md:28]: Fixed — updated to `mst_abc123` to match the `MemoryStoreIdMarker::PREFIX`.

File: specs/memory.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+  "type": "object",
+  "properties": {
+    "content": { "type": "string", "maxLength": 2000, "description": "1-3 sentence knowledge to persist" },
+    "kind": { "type": "string", "enum": ["fact", "preference", "correction", "procedure", "context"], "default": "fact" },
+    "importance": { "type": "integer", "minimum": 1, "maximum": 10, "default": 5 },
+    "tags": { "type": "array", "items": { "type": "string" }, "maxItems": 10 }
+  },
+  "required": ["content"],
+  "additionalProperties": false
+}
```
[the remark this answers is not in the document]
Comment (chaliy) [reply specs/memory.md:57]: Fixed — added the `images` field with `base64`/`media_type` object items and `maxItems: 4` to the spec schema.

File: specs/memory.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+  "memories": [
+    {
+      "id": "mem_xxx",
+      "kind": "correction",
+      "importance": 8,
+      "created_at": "2026-03-24T12:00:00Z",
+      "content_parts": [
+        { "type": "text", "text": "Don't use unwrap() in production code..." },
+        { "type": "image", "url": "data:image/png;base64,...", "media_type": "image/png" }
+      ],
```
[the remark this answers is not in the document]
Comment (chaliy) [reply specs/memory.md:92]: Fixed — updated the example to use `{ "type": "image", "base64": "iVBORw0KGgo=...", "media_type": "image/png" }` matching the actual `MemoryImagePart` shape.

File: crates/core/src/capabilities/persistent_memory.rs [hunk 4]
```diff
@@ -0,0 +1,854 @@
...
+            .kind
+            .as_deref()
+            .and_then(|k| serde_json::from_value(json!(k)).ok())
+            .unwrap_or(MemoryKind::Fact);
+
+        // Parse config from empty (tools don't have direct config access, use defaults)
+        let config = MemoryConfig::default();
+        let store_id = match resolve_store_id(&config, context).await {
+            Ok(id) => id,
+            Err(e) => return e,
```
[the remark this answers is not in the document]
Comment (chaliy) [reply crates/core/src/capabilities/persistent_memory.rs]: Agreed — tools don't currently have access to per-agent `AgentCapabilityConfig`. Added a `TODO(V1)` comment at all three tool sites documenting this as a known limitation. Resolving config from capability config requires extending `ToolContext` with the agent's capability config, which is a cross-cutting change better suited for a follow-up.

File: crates/core/src/capabilities/persistent_memory.rs [hunk 5]
```diff
@@ -0,0 +1,854 @@
...
+        if let Err(e) = MemoryLimits::validate(&params.content, &params.tags, &content_parts) {
+            return ToolExecutionResult::tool_error(e);
+        }
+
+        let kind = params
+            .kind
+            .as_deref()
+            .and_then(|k| serde_json::from_value(json!(k)).ok())
+            .unwrap_or(MemoryKind::Fact);
+
```
[the remark this answers is not in the document]
Comment (chaliy) [reply crates/core/src/capabilities/persistent_memory.rs]: Fixed — both `remember` and `recall` now return a `ToolError` with a descriptive message listing valid values when an invalid `kind` is provided, instead of silently defaulting.

File: crates/core/src/capabilities/persistent_memory.rs [hunk 6]
```diff
@@ -0,0 +1,854 @@
...
+        let backend = match context.memory_store.as_ref() {
+            Some(b) => b,
+            None => return ToolExecutionResult::tool_error("Memory store not available"),
+        };
+
+        match backend.forget(memory_id).await {
+            Ok(true) => ToolExecutionResult::success(json!({
+                "forgotten": true,
+                "memory_id": params.memory_id
```
[the remark this answers is not in the document]
Comment (chaliy) [reply crates/core/src/capabilities/persistent_memory.rs:582]: Fixed — `forget` now resolves the store ID (via the same `resolve_store_id` path as remember/recall) and passes `store_id` to the backend. The `MemoryStoreBackend::forget` trait method signature now requires `store_id`, and the `InMemoryMemoryStore` implementation enforces that the memory belongs to the given store before deactivating.

File: crates/core/src/memory_store.rs [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+        importance: u8,
+        tags: Vec<String>,
+    ) -> crate::error::Result<Memory>;
+
+    /// Search/recall memories.
+    async fn recall(&self, query: MemoryQuery) -> crate::error::Result<(Vec<Memory>, usize)>;
+
+    /// Soft-delete (deactivate) a memory.
+    async fn forget(&self, memory_id: MemoryId) -> crate::error::Result<bool>;
+
```
[the remark this answers is not in the document]
Comment (chaliy) [reply crates/core/src/memory_store.rs:267]: Fixed — `forget()` now takes `store_id` as a required parameter. Combined with the `resolve_store_id` changes (which verify org ownership), this ensures tenant isolation at the API boundary. `create_memory` already takes `store_id`, and org ownership is now verified before the store_id reaches these methods.

File: crates/core/src/traits.rs [hunk 8]
```diff
@@ -532,6 +532,12 @@ pub struct ToolContext {
     pub tool_call_id: Option<String>,
     /// Optional capability registry for blueprint lookups.
     pub capability_registry: Option<crate::capabilities::CapabilityRegistry>,
+
+    /// Optional memory store backend for persistent cross-session memory.
+    pub memory_store: Option<Arc<dyn crate::memory_store::MemoryStoreBackend>>,
+
+    /// Optional org ID for org-scoped operations (memory stores, etc.).
+    pub org_id: Option<crate::typed_id::OrgId>,
```
[the remark this answers is not in the document]
Comment (chaliy) [reply crates/core/src/traits.rs:540]: Fixed — added `memory_store` (as `is_some()` bool) and `org_id` to the `Debug` output.

File: specs/memory.md [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+- `MemoryStore` — org-scoped named store with capacity config
+- `Memory` — content, kind, importance, tags, active flag, store FK
+- `MemoryContentPart` — text or image (same shape as `ContentPart`, minus tool variants)
+
+## Capacity Limits
+
+| Limit | Default | Scope | Notes |
+|-------|---------|-------|-------|
+| Active memories per store | 10,000 | Per-store | Oldest low-importance auto-archived beyond this |
+| Stores per org | 50 | Per-org | |
```
[the remark this answers is not in the document]
Comment (chaliy) [reply specs/memory.md:131]: Fixed — updated the spec to say "Hard cap; `remember` returns error when full" instead of "auto-archived", matching the actual implementation.

File: crates/core/src/capabilities/persistent_memory.rs [hunk 10]
```diff
@@ -0,0 +1,854 @@
...
+    let backend = context
+        .memory_store
+        .as_ref()
+        .ok_or_else(|| ToolExecutionResult::tool_error("Memory store not available"))?;
+
+    if let Some(ref store_str) = config.store {
+        store_str
+            .parse::<MemoryStoreId>()
+            .map_err(|_| ToolExecutionResult::tool_error(format!("Invalid store ID: {store_str}")))
+    } else {
```
[the remark this answers is not in the document]
Comment (chaliy) [reply crates/core/src/capabilities/persistent_memory.rs:140]: Fixed — `resolve_store_id` now fetches the store via `get_store`, verifies it exists, and checks `store.org_id == context.org_id` before returning. Returns a generic "not found" error on mismatch to avoid leaking store existence across orgs.

File: crates/core/src/capabilities/persistent_memory.rs [hunk 11]
```diff
@@ -0,0 +1,854 @@
...
+                params.importance,
+                params.tags,
+            )
+            .await
+        {
+            Ok(memory) => ToolExecutionResult::success(json!({
+                "memory_id": memory.id.to_string(),
+                "created": true
+            })),
+            Err(e) => ToolExecutionResult::tool_error(format!("Failed to save memory: {e}")),
```
[the remark this answers is not in the document]
Comment (chaliy) [reply crates/core/src/capabilities/persistent_memory.rs:331]: Fixed — updated spec and tool description to reflect create-only semantics for V1. Removed "deduplicated on write" and "Create or update" language, removed the `created: false` merge response variant. Dedup/merge is a natural follow-up once we have embedding-based similarity search.
